### PR TITLE
Replaced coingeko with ms-rates

### PR DESCRIPTION
**Pull request by MayukhSamadder2000**
Replace the native asset rate and the asset by token address rate from coingeko with a new function from the sdk that uses the fetchExchangeRates

### CODE REVIEW

File: src/providers/EtherspotContextProvider.tsx [hunk 1]
```diff
@@ -378,6 +379,38 @@ const EtherspotContextProvider = ({
     [sdk, accountAddress]
   );
 
+  const getRatesByTokenAddresses = useCallback(async (chainId: number, tokenAddresses: string[]) => {
+    let tokens  = tokenAddresses.filter((address) => address !== null)
+    if(!sdk || !tokens.length) return null
```
Comment (Yash-Chaubal) [inline src/providers/EtherspotContextProvider.tsx:384]:
```suggestion
    if (!sdk || !tokens.length) return null
```

File: src/providers/EtherspotContextProvider.tsx [hunk 2]
```diff
@@ -378,6 +379,38 @@ const EtherspotContextProvider = ({
...
+    try {
+      const rates: RateData = await sdk.fetchExchangeRates({tokens, chainId});
+      if(!rates.errored && rates.items.length > 0) {
```
Comment (Yash-Chaubal) [inline src/providers/EtherspotContextProvider.tsx:387]: Maybe just add a few null checks considering we're interacting with the sdk and don't know if something will be returned
```suggestion
      if (!rates?.errored && rates?.items?.length) {
```

File: src/providers/EtherspotContextProvider.tsx [hunk 3]
```diff
@@ -378,6 +379,38 @@ const EtherspotContextProvider = ({
...
+        return addressByChain
+      }
+    } catch (error) {
+      console.log("rateserror",error)
+    }
+
+  }, [sdk])
+
+  const getRatesByNativeChainId = useCallback(async (chainId: number) => {
+    if(!sdk) return null
```
Comment (Yash-Chaubal) [inline src/providers/EtherspotContextProvider.tsx:401]:
```suggestion
    if (!sdk) return null
```

File: src/providers/EtherspotContextProvider.tsx [hunk 4]
```diff
@@ -378,6 +379,38 @@ const EtherspotContextProvider = ({
...
+    try {
+      const rates: RateData = await sdk.fetchExchangeRates({tokens: ['0x0000000000000000000000000000000000000000'], chainId});
+      if(!rates.errored && rates.items.length > 0) {
```
Comment (Yash-Chaubal) [inline src/providers/EtherspotContextProvider.tsx:404]: Same as above
```suggestion
      if (!rates?.errored && rates?.items?.length) {
```

File: src/providers/EtherspotContextProvider.tsx [hunk 5]
```diff
@@ -378,6 +379,38 @@ const EtherspotContextProvider = ({
...
+    let tokens  = tokenAddresses.filter((address) => address !== null)
+    if(!sdk || !tokens.length) return null
+    try {
+      const rates: RateData = await sdk.fetchExchangeRates({tokens, chainId});
+      if(!rates.errored && rates.items.length > 0) {
+        let addressByChain : Record<string, number> = {}
+        rates.items.forEach((rate) => {
+          addressByChain[rate.address] = rate.usd
+        })
+        return addressByChain
```
Comment (Yash-Chaubal) [inline src/providers/EtherspotContextProvider.tsx:392]: Should we change the var name `addressByChain`? It doesn't really represent what's being returned

**Review by Yash-Chaubal — APPROVED** [review]
LGTM! Just left a few comments - are you using prettier as a default formatter and formatting on save?

File: src/providers/EtherspotContextProvider.tsx (hunk 2 above)
Comment (MayukhSamadder2000) [reply src/providers/EtherspotContextProvider.tsx:387]: @Yash-Chaubal so we get this key everytime the requests succeed so rates.errored is always there if you see the Type `RateData`

File: src/providers/EtherspotContextProvider.tsx (hunk 4 above)
Comment (MayukhSamadder2000) [reply src/providers/EtherspotContextProvider.tsx:404]: @Yash-Chaubal so we get this key everytime the requests succeed so rates.errored is always there if you see the Type RateData

File: src/providers/EtherspotContextProvider.tsx (hunk 5 above)
Comment (MayukhSamadder2000) [reply src/providers/EtherspotContextProvider.tsx]: nah I guess it should be something like `rateByAddress`

File: src/providers/EtherspotContextProvider.tsx [hunk 6]
```diff
@@ -364,6 +364,41 @@ const EtherspotContextProvider = ({
     [sdk, accountAddress]
   );
 
+  const getRatesByTokenAddresses = useCallback(
```
Comment (poocart) [inline src/providers/EtherspotContextProvider.tsx:367]: can be without `useCallback`

File: src/providers/EtherspotContextProvider.tsx [hunk 7]
```diff
@@ -364,6 +364,41 @@ const EtherspotContextProvider = ({
...
+    },
+    [sdk]
+  );
+
+  const getRatesByNativeChainId = useCallback(
+    async (chainId: number) => {
+      if (!sdk) return null;
+      try {
+        const rates: RateData = await sdk.fetchExchangeRates({
+          tokens: ['0x0000000000000000000000000000000000000000'],
```
Comment (poocart) [inline src/providers/EtherspotContextProvider.tsx:390]: is this right and works for all native tokens?

Comment (poocart) [reply src/providers/EtherspotContextProvider.tsx:390]: also 0x can be replaced by zero address constant

File: src/providers/EtherspotContextProvider.tsx [hunk 8]
```diff
@@ -364,6 +364,41 @@ const EtherspotContextProvider = ({
     [sdk, accountAddress]
   );
 
+  const getRatesByTokenAddresses = useCallback(
+    async (chainId: number, tokenAddresses: string[]) => {
+      let tokens = tokenAddresses.filter((address) => address !== null);
```
Comment (poocart) [inline src/providers/EtherspotContextProvider.tsx:369]:
```suggestion
      const tokens = tokenAddresses.filter((address) => address !== null);
```

Comment (poocart) [reply src/providers/EtherspotContextProvider.tsx:369]: also consider adding zero address checks, not just null

File: src/providers/EtherspotContextProvider.tsx [hunk 9]
```diff
@@ -364,6 +364,41 @@ const EtherspotContextProvider = ({
...
+      if (!sdk || !tokens.length) return null;
+      try {
+        const rates: RateData = await sdk.fetchExchangeRates({ tokens, chainId });
+        if (!rates.errored && rates.items.length) {
+          let rateByAddress: Record<string, number> = {};
+          rates.items.forEach((rate) => {
+            rateByAddress[rate.address] = rate.usd;
+          });
+          return rateByAddress;
+        }
```
Comment (poocart) [inline src/providers/EtherspotContextProvider.tsx:379]: can be just

if (rates.errored || !rates?.items?.length) return null;

then no extra if wrap

File: src/providers/EtherspotContextProvider.tsx (hunk 6 above)
Comment (MayukhSamadder2000) [reply src/providers/EtherspotContextProvider.tsx:367]: okh

File: src/providers/EtherspotContextProvider.tsx [hunk 10]
```diff
@@ -364,6 +364,41 @@ const EtherspotContextProvider = ({
...
+      } catch (error) {}
+    },
+    [sdk]
+  );
+
+  const getRatesByNativeChainId = useCallback(
```
[the remark this answers is not in the document]
Comment (MayukhSamadder2000) [reply src/providers/EtherspotContextProvider.tsx:385]: okh

File: src/providers/EtherspotContextProvider.tsx (hunk 7 above)
Comment (MayukhSamadder2000) [reply src/providers/EtherspotContextProvider.tsx:390]: Yeah so I asked Vignesh for the native token thing he said 0x0000000000000000000000000000000000000000 will do

File: src/providers/EtherspotContextProvider.tsx (hunk 8 above)
Comment (MayukhSamadder2000) [reply src/providers/EtherspotContextProvider.tsx:369]: I see this variable called `zeroAddressCinstants` it has a list of addresses

Comment (MayukhSamadder2000) [reply src/providers/EtherspotContextProvider.tsx:369]: i think we can use the function `isZeroAddress`

File: src/providers/EtherspotContextProvider.tsx [hunk 11]
```diff
@@ -364,6 +364,35 @@ const EtherspotContextProvider = ({
...
+    if (!sdk || !tokens.length) return null;
+    try {
+      const rates: RateData = await sdk.fetchExchangeRates({ tokens, chainId });
+      if (rates.errored || !rates?.items?.length) return null;
+
+      let rateByAddress: Record<string, number> = {};
+      rates.items.forEach((rate) => {
+        rateByAddress[rate.address] = rate.usd;
+      });
+      return rateByAddress;
```
Comment (poocart) [inline src/providers/EtherspotContextProvider.tsx:378]:
```suggestion
      return rates.items.reduce((rateByAddress, rate) => ({
        ...rateByAddress,
        [rate.address]: rate.usd,
      }));
```

File: src/providers/EtherspotContextProvider.tsx [hunk 12]
```diff
@@ -364,6 +364,35 @@ const EtherspotContextProvider = ({
...
+    } catch (error) {}
```
Comment (poocart) [inline src/providers/EtherspotContextProvider.tsx:379]:
```suggestion
    } catch (error) {
      //
    }
```

File: src/providers/EtherspotContextProvider.tsx [hunk 13]
```diff
@@ -364,6 +364,35 @@ const EtherspotContextProvider = ({
     [sdk, accountAddress]
   );
 
+  const getRatesByTokenAddresses = async (chainId: number, tokenAddresses: string[]) => {
+    const tokens = tokenAddresses.filter((address) => address !== null && !isZeroAddress(address));
+    if (!sdk || !tokens.length) return null;
+    try {
```
Comment (poocart) [inline src/providers/EtherspotContextProvider.tsx:370]:
```suggestion
    if (!sdk || !tokens.length) return null;

    try {
```

File: src/providers/EtherspotContextProvider.tsx [hunk 14]
```diff
@@ -364,6 +364,35 @@ const EtherspotContextProvider = ({
...
+      rates.items.forEach((rate) => {
+        rateByAddress[rate.address] = rate.usd;
+      });
+      return rateByAddress;
+    } catch (error) {}
+  };
+
+  const getRatesByNativeChainId = async (chainId: number) => {
+    if (!sdk) return null;
+    try {
```
Comment (poocart) [inline src/providers/EtherspotContextProvider.tsx:384]:
```suggestion
    if (!sdk) return null;
 
    try {
```

File: src/providers/EtherspotContextProvider.tsx [hunk 15]
```diff
@@ -364,6 +364,35 @@ const EtherspotContextProvider = ({
...
+      const rates: RateData = await sdk.fetchExchangeRates({
+        tokens: [ethers.constants.AddressZero],
+        chainId,
+      });
+      if (!rates.errored && rates.items.length) {
+        return rates.items[0].usd;
+      }
+    } catch (error) {}
+    return null;
```
Comment (poocart) [inline src/providers/EtherspotContextProvider.tsx:393]:
```suggestion
    } catch (error) {
      //
    }

    return null;
```

**Pull request merged**